http-validation/src/main/java/io/micronaut/validation/routes/FilterVisitor.java

```diff
@@ -0,0 +1,130 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.validation.routes;
+
+import io.micronaut.core.annotation.AnnotationValue;
+import io.micronaut.core.async.publisher.Publishers;
+import io.micronaut.core.type.Argument;
+import io.micronaut.http.HttpRequest;
+import io.micronaut.http.HttpResponse;
+import io.micronaut.http.MutableHttpRequest;
+import io.micronaut.http.MutableHttpResponse;
+import io.micronaut.http.annotation.ClientFilter;
+import io.micronaut.http.annotation.RequestFilter;
+import io.micronaut.http.annotation.ResponseFilter;
+import io.micronaut.http.annotation.ServerFilter;
+import io.micronaut.http.filter.FilterContinuation;
+import io.micronaut.http.filter.FilterRunner;
+import io.micronaut.inject.ast.ClassElement;
+import io.micronaut.inject.ast.MethodElement;
+import io.micronaut.inject.ast.ParameterElement;
+import io.micronaut.inject.visitor.TypeElementVisitor;
+import io.micronaut.inject.visitor.VisitorContext;
+import org.reactivestreams.Publisher;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * Visitor that checks validity of {@link ServerFilter} and {@link ClientFilter} classes.
+ */
+public final class FilterVisitor implements TypeElementVisitor<Object, Object> {
+    private static final Set<Class<?>> PERMITTED_CLASSES = Set.of(
+        void.class,
+        HttpRequest.class,
+        MutableHttpRequest.class,
+        HttpResponse.class,
+        MutableHttpResponse.class,
+        FilterContinuation.class
+    );
+
+    @Override
+    public Set<String> getSupportedAnnotationNames() {
+        return Set.of(
+            RequestFilter.class.getName(),
+            ResponseFilter.class.getName()
+        );
+    }
+
+    @Override
+    public void visitMethod(MethodElement element, VisitorContext context) {
+        AnnotationValue<RequestFilter> requestFilterAnnotation = element.getAnnotation(RequestFilter.class);
+        AnnotationValue<ResponseFilter> responseFilterAnnotation = element.getAnnotation(ResponseFilter.class);
+        if (requestFilterAnnotation == null && responseFilterAnnotation == null) {
+            return;
+        }
+
+        if (!element.getDeclaringType().isAnnotationPresent(ServerFilter.class) &&
+            !element.getDeclaringType().isAnnotationPresent(ClientFilter.class)) {
+
+            context.fail("Filter method must be declared on a @ServerFilter or @ClientFilter bean", element);
+            return;
+        }
+
+        try {
+            Argument<?>[] args = toArguments(Arrays.stream(element.getParameters()).map(ParameterElement::getType).toList(), context);
+            Argument<?> ret = toArgument(element.getReturnType(), context);
+
+            if (requestFilterAnnotation != null) {
+                FilterRunner.prepareFilterMethod(null, null, args, ret, false, null);
```

Comment:
it is not clear to me what this call is doing and why it is in a source code processor. You seem to be converting the source code elements to runtime types? And then what?

Reply:
it will throw an exception when the filter is invalid. i added a comment. it's a bit weird to mix compile time and run time types like this, and it's right on the edge of "too smart to be understandable" imo. But i think it makes the code genuinely simpler and reduces duplication.